patch_rebaser/patch_rebaser.py

```diff
@@ -109,6 +111,96 @@ def set_up_git_config(name, email):
+        """Initialise the Rebaser
+
+       :param git_wrapper.GitRepo repo: An initialised GitWrapper repo
+       :param str branch: Local branch name to rebase
+       :param str commit: Commit sha to rebase to
+       :param str remote: Remote name
+       :param str tag_name: Tag name to push to the remote
+       :param bool dev_mode: Whether to run the push commands as dry-run only
+       :param int max_retries: How many retry attempts if remote changed during
+                               rebase
+        """
+        self.repo = repo
+        self.branch = branch
+        self.commit = commit
+        self.remote = remote
+        self.tag_name = tag_name
+        self.dev_mode = dev_mode
+        self.max_retries = max_retries
+
+        self.remote_branch = "{0}/{1}".format(self.remote, self.branch)
+
+    def rebase_and_update_remote(self):
+        """Rebase the local branch to the specific commit & push the result."""
+        self.repo.remote.fetch(self.remote)
+
+        # Reset the local branch to the latest
+        self.repo.branch.create(self.branch,
+                                self.remote_branch,
+                                reset_if_exists=True)
+
+        # Tag the previous branch tip, before rebase
+        tag = self.repo.repo.create_tag(self.tag_name, self.remote_branch)
+
+        # Rebase
+        self.perform_rebase()
+
+        # Check if any new changes have come in
+        self.repo.remote.fetch(self.remote)
+
+        # TODO: Update git_wrapper to offer a more sensible comparison function
+        remote_tip_ref = self.repo.repo.remotes[self.remote].refs[self.branch]
+        remote_sha = remote_tip_ref.commit.hexsha
+        if tag.commit.hexsha != remote_sha:
+            if self.max_retries > 0:
+                LOGGER.info("Remote changed during rebase. Remaining "
+                            "attempts: %s", self.max_retries)
+                time.sleep(20)
+                self.max_retries -= 1
+
+                # We'll need to move the tag to the new HEAD
+                self.repo.git.tag("-d", self.tag_name)
+                self.rebase_and_update_remote()
+            else:
+                LOGGER.warning("Remote changed multiple times during rebase, "
+                               "not pushing.")
+        else:
+            # No new stuff, push it on
+            self.update_remote_patches_branch()
+
+    def perform_rebase(self):
+        """Rebase the specific local branch to the specific commit."""
+        try:
+            LOGGER.info("Rebasing %s to %s", self.branch, self.commit)
+            self.repo.branch.rebase_to_hash(self.branch, self.commit)
+        except git_exceptions.RebaseException:
+            LOGGER.info("Could not rebase. Cleaning up.")
+            self.repo.branch.abort_rebase()
+            raise
+
+    def update_remote_patches_branch(self):
+        """Force push local patches branch to the remote repository."""
```

Comment:
Are we sure that we need a force push, this should be a standard rebase that does not require a force push. I would recommend that we not do a force and let things start to fail. I would rather a human do a force push instead of automation.

Reply:
It should not need to force push in this case. This would be similar to rebasing a development branch with master, the commits will get new hashes but you are not modifying the content of the commits themselves so a force should not be needed.

I would recommend that we disable force push and see if things fail and re-evaluate at that point. In my experience force push is rarely needed.

Reply:
After doing some more thinking and some experimentation, we have no choice but to use force push. I would recommend that we make sure to log that we are doing a force push with a timestamp so we can look back in logs if something bad happens.